File: 0056-merge-intervals/0056-merge-intervals.py
class Solution:
    def merge(self, intervals: List[List[int]]) -> List[List[int]]:
        # Base case 
        if len(intervals) == 1: 
            return intervals
        
        # Sort intervals by end time, ascending order
        intervals.sort(key=lambda x: x[0])
        
        # Declare variables result array and initiate start and end to be first interval
        result = []
        start = intervals[0][0]
        end = intervals[0][1]
        
        for i in range(1, len(intervals)):
            # Destructure current interval
            currStart, currEnd = intervals[i][0], intervals[i][1]
            
            if currStart <= end:
                end = max(end, currEnd)
            else:
                result.append([start, end])
                start = currStart
                end = currEnd
        
        result.append([start, end])
        return result
        
        

File: 0056-merge-intervals/0056-merge-intervals.js
/**
 * @param {number[][]} intervals
 * @return {number[][]}
 */
var merge = function(intervals) {
  // Base case
  if (intervals.length === 1) return intervals;
  
  // Sort start times
  intervals.sort((a,b) => a[0] - b[0]);
  
  // Declare variables start, end and result
  const result = [];
  let start = intervals[0][0];
  let end = intervals[0][1];
  
  // Iterate through intervals to merge intervals by reassigning start and end
  for (let i = 1; i < intervals.length; i++) {
    const [currStart, currEnd] = intervals[i];
    
    if (end >= currStart) {
      end = Math.max(end, currEnd);
    } else {
      result.push([start, end]);
      start = currStart;
      end = currEnd;
    }
  }
  
  result.push([start, end]);
  
  return result;
};